File: kayak-backend/src/models/entities/device.rs
//! 设备实体模型
//!
//! 定义设备表的数据结构和相关枚举

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// 协议类型枚举
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProtocolType {
    /// 虚拟设备
    Virtual,
    /// Modbus TCP
    ModbusTcp,
    /// Modbus RTU
    ModbusRtu,
    /// CAN总线
    Can,
    /// VISA协议
    Visa,
    /// MQTT
    Mqtt,
}

/// 设备状态枚举
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum DeviceStatus {
    /// 离线
    #[default]
    Offline,
    /// 在线
    Online,
    /// 错误
    Error,
}

/// 设备实体
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Device {
    /// 设备ID (UUID)
    pub id: Uuid,
    /// 所属工作台ID
    pub workbench_id: Uuid,
    /// 父设备ID (支持嵌套)
    pub parent_id: Option<Uuid>,
    /// 设备名称
    pub name: String,
    /// 协议类型
    pub protocol_type: ProtocolType,
    /// 协议参数 (JSON格式)
    pub protocol_params: Option<serde_json::Value>,
    /// 制造商
    pub manufacturer: Option<String>,
    /// 型号
    pub model: Option<String>,
    /// 序列号
    pub sn: Option<String>,
    /// 状态
    pub status: DeviceStatus,
    /// 创建时间
    pub created_at: DateTime<Utc>,
    /// 更新时间
    pub updated_at: DateTime<Utc>,
}

impl Device {
    /// 创建设备
    pub fn new(
        workbench_id: Uuid,
        name: String,
        protocol_type: ProtocolType,
        parent_id: Option<Uuid>,
    ) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::new_v4(),
            workbench_id,
            parent_id,
            name,
            protocol_type,
            protocol_params: None,
            manufacturer: None,
            model: None,
            sn: None,
            status: DeviceStatus::Offline,
            created_at: now,
            updated_at: now,
        }
    }
}

/// 创建设备请求DTO
#[derive(Debug, Deserialize)]
pub struct CreateDeviceRequest {
    pub workbench_id: Uuid,
    pub name: String,
    pub protocol_type: ProtocolType,
    pub parent_id: Option<Uuid>,
    pub protocol_params: Option<serde_json::Value>,
    pub manufacturer: Option<String>,
    pub model: Option<String>,
    pub sn: Option<String>,
}

/// 更新设备请求DTO
#[derive(Debug, Deserialize, Default)]
pub struct UpdateDeviceRequest {
    pub name: Option<String>,
    pub protocol_params: Option<serde_json::Value>,
    pub manufacturer: Option<String>,
    pub model: Option<String>,
    pub sn: Option<String>,
    pub status: Option<DeviceStatus>,
}

/// 设备响应DTO
#[derive(Debug, Serialize)]
pub struct DeviceResponse {
    pub id: Uuid,
    pub workbench_id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub protocol_type: ProtocolType,
    pub protocol_params: Option<serde_json::Value>,
    pub manufacturer: Option<String>,
    pub model: Option<String>,
    pub sn: Option<String>,
    pub status: DeviceStatus,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<Device> for DeviceResponse {
    fn from(device: Device) -> Self {
        Self {
            id: device.id,
            workbench_id: device.workbench_id,
            parent_id: device.parent_id,
            name: device.name,
            protocol_type: device.protocol_type,
            protocol_params: device.protocol_params,
            manufacturer: device.manufacturer,
            model: device.model,
            sn: device.sn,
            status: device.status,
            created_at: device.created_at,
            updated_at: device.updated_at,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // TC-S1-016-01: 创建设备（无父设备）测试
    #[test]
    fn test_create_device_without_parent() {
        let workbench_id = Uuid::new_v4();
        let device = Device::new(
            workbench_id,
            "Test Device".to_string(),
            ProtocolType::Virtual,
            None,
        );

        assert!(device.parent_id.is_none());
        assert_eq!(device.status, DeviceStatus::Offline);
        assert_eq!(device.name, "Test Device");
        assert_eq!(device.protocol_type, ProtocolType::Virtual);
    }

    // TC-S1-016-02: 创建设备（带父设备）测试
    #[test]
    fn test_create_device_with_parent() {
        let workbench_id = Uuid::new_v4();
        let parent_id = Uuid::new_v4();
        let device = Device::new(
            workbench_id,
            "Child Device".to_string(),
            ProtocolType::ModbusTcp,
            Some(parent_id),
        );

        assert_eq!(device.parent_id, Some(parent_id));
        assert_ne!(device.id, parent_id);
    }

    // TC-S1-016-03: 验证UUID唯一性
    #[test]
    fn test_device_uuid_uniqueness() {
        let workbench_id = Uuid::new_v4();
        let device1 = Device::new(
            workbench_id,
            "Device 1".to_string(),
            ProtocolType::Virtual,
            None,
        );
        let device2 = Device::new(
            workbench_id,
            "Device 2".to_string(),
            ProtocolType::Virtual,
            None,
        );
        assert_ne!(device1.id, device2.id);
    }

    // TC-S1-016-05: 设备树形结构（三层）测试
    #[test]
    fn test_device_tree_structure() {
        let workbench_id = Uuid::new_v4();
        let grandparent = Device::new(
            workbench_id,
            "Grandparent".to_string(),
            ProtocolType::Virtual,
            None,
        );
        let parent = Device::new(
            workbench_id,
            "Parent".to_string(),
            ProtocolType::Virtual,
            Some(grandparent.id),
        );
        let child = Device::new(
            workbench_id,
            "Child".to_string(),
            ProtocolType::Virtual,
            Some(parent.id),
        );

        assert!(grandparent.parent_id.is_none());
        assert_eq!(parent.parent_id, Some(grandparent.id));
        assert_eq!(child.parent_id, Some(parent.id));
    }

    // TC-S1-016-08: ProtocolType 枚举所有变体测试
    #[test]
    fn test_protocol_type_variants() {
        let workbench_id = Uuid::new_v4();
        for protocol in &[
            ProtocolType::Virtual,
            ProtocolType::ModbusTcp,
            ProtocolType::ModbusRtu,
            ProtocolType::Can,
            ProtocolType::Visa,
            ProtocolType::Mqtt,
        ] {
            let device = Device::new(workbench_id, format!("{:?}", protocol), *protocol, None);
            assert_eq!(device.protocol_type, *protocol);
        }
    }

    // TC-S1-016-23: Device 转 DeviceResponse 测试
    #[test]
    fn test_device_to_response() {
        let workbench_id = Uuid::new_v4();
        let device = Device::new(
            workbench_id,
            "Test Device".to_string(),
            ProtocolType::Virtual,
            None,
        );
        let response = DeviceResponse::from(device.clone());

        assert_eq!(response.id, device.id);
        assert_eq!(response.workbench_id, device.workbench_id);
        assert_eq!(response.name, device.name);
        assert_eq!(response.protocol_type, device.protocol_type);
        assert_eq!(response.status, device.status);
        assert_eq!(response.created_at, device.created_at);
        assert_eq!(response.updated_at, device.updated_at);
    }

    // TC-S1-016-27: ProtocolType JSON 序列化测试
    #[test]
    fn test_protocol_type_serialization() {
        let json = serde_json::to_string(&ProtocolType::Virtual).unwrap();
        assert_eq!(json, "\"virtual\"");
        let json = serde_json::to_string(&ProtocolType::ModbusTcp).unwrap();
        assert_eq!(json, "\"modbus_tcp\"");
    }

    // TC-S1-016-30: JSON 反序列化测试
    #[test]
    fn test_protocol_type_deserialization() {
        let protocol: ProtocolType = serde_json::from_str("\"virtual\"").unwrap();
        assert_eq!(protocol, ProtocolType::Virtual);
    }

    // DeviceStatus 反序列化测试
    #[test]
    fn test_device_status_deserialization() {
        let status1: DeviceStatus = serde_json::from_str("\"online\"").unwrap();
        assert_eq!(status1, DeviceStatus::Online);
        let status2: DeviceStatus = serde_json::from_str("\"error\"").unwrap();
        assert_eq!(status2, DeviceStatus::Error);
        let status3: DeviceStatus = serde_json::from_str("\"offline\"").unwrap();
        assert_eq!(status3, DeviceStatus::Offline);
    }

    // TC-S1-016-33: 设备可选字段（带值）测试
    #[test]
    fn test_device_optional_fields_with_values() {
        let workbench_id = Uuid::new_v4();
        let mut device = Device::new(
            workbench_id,
            "Test".to_string(),
            ProtocolType::ModbusTcp,
            None,
        );
        device.manufacturer = Some("Acme Corp".to_string());
        device.model = Some("Model-X".to_string());
        device.sn = Some("SN12345".to_string());
        device.protocol_params = Some(serde_json::json!({"host": "192.168.1.1", "port": 502}));

        assert_eq!(device.manufacturer, Some("Acme Corp".to_string()));
        assert_eq!(device.model, Some("Model-X".to_string()));
        assert_eq!(device.sn, Some("SN12345".to_string()));
        assert!(device.protocol_params.is_some());
    }
}

File: kayak-frontend/lib/models/device.dart
import 'package:json_annotation/json_annotation.dart';

part 'device.g.dart';

// ============================================================
// ProtocolType — 协议类型枚举
// ============================================================
enum ProtocolType {
  @JsonValue('virtual')
  virtual,
  @JsonValue('modbus_tcp')
  modbusTcp,
  @JsonValue('modbus_rtu')
  modbusRtu,
  @JsonValue('can')
  can,
  @JsonValue('visa')
  visa,
  @JsonValue('mqtt')
  mqtt,
}

// ============================================================
// Device — 设备实体
// ============================================================
@JsonSerializable()
class Device {

  const Device({
    required this.id,
    required this.workbenchId,
    this.parentId,
    required this.name,
    required this.protocolType,
    this.protocolParams,
    this.manufacturer,
    this.model,
    this.sn,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Device.fromJson(Map<String, dynamic> json) => _$DeviceFromJson(json);
  final String id;
  @JsonKey(name: 'workbench_id')
  final String workbenchId;
  @JsonKey(name: 'parent_id')
  final String? parentId;
  final String name;
  @JsonKey(name: 'protocol_type')
  final ProtocolType protocolType;
  @JsonKey(name: 'protocol_params')
  final Map<String, dynamic>? protocolParams;
  final String? manufacturer;
  final String? model;
  final String? sn;
  final String status;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;
  Map<String, dynamic> toJson() => _$DeviceToJson(this);

  Device copyWith({
    String? id,
    String? workbenchId,
    String? parentId,
    String? name,
    ProtocolType? protocolType,
    Map<String, dynamic>? protocolParams,
    String? manufacturer,
    String? model,
    String? sn,
    String? status,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Device(
      id: id ?? this.id,
      workbenchId: workbenchId ?? this.workbenchId,
      parentId: parentId ?? this.parentId,
      name: name ?? this.name,
      protocolType: protocolType ?? this.protocolType,
      protocolParams: protocolParams ?? this.protocolParams,
      manufacturer: manufacturer ?? this.manufacturer,
      model: model ?? this.model,
      sn: sn ?? this.sn,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

// ============================================================
// DeviceTreeNode — 设备树节点（递归结构）
// ============================================================
@JsonSerializable()
class DeviceTreeNode {

  const DeviceTreeNode({
    required this.id,
    required this.workbenchId,
    this.parentId,
    required this.name,
    required this.protocolType,
    this.protocolParams,
    this.manufacturer,
    this.model,
    this.sn,
    required this.status,
    this.children = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory DeviceTreeNode.fromJson(Map<String, dynamic> json) =>
      _$DeviceTreeNodeFromJson(json);
  final String id;
  @JsonKey(name: 'workbench_id')
  final String workbenchId;
  @JsonKey(name: 'parent_id')
  final String? parentId;
  final String name;
  @JsonKey(name: 'protocol_type')
  final ProtocolType protocolType;
  @JsonKey(name: 'protocol_params')
  final Map<String, dynamic>? protocolParams;
  final String? manufacturer;
  final String? model;
  final String? sn;
  final String status;
  @JsonKey(defaultValue: [])
  final List<DeviceTreeNode> children;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;
  Map<String, dynamic> toJson() => _$DeviceTreeNodeToJson(this);

  DeviceTreeNode copyWith({
    String? id,
    String? workbenchId,
    String? parentId,
    String? name,
    ProtocolType? protocolType,
    Map<String, dynamic>? protocolParams,
    String? manufacturer,
    String? model,
    String? sn,
    String? status,
    List<DeviceTreeNode>? children,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return DeviceTreeNode(
      id: id ?? this.id,
      workbenchId: workbenchId ?? this.workbenchId,
      parentId: parentId ?? this.parentId,
      name: name ?? this.name,
      protocolType: protocolType ?? this.protocolType,
      protocolParams: protocolParams ?? this.protocolParams,
      manufacturer: manufacturer ?? this.manufacturer,
      model: model ?? this.model,
      sn: sn ?? this.sn,
      status: status ?? this.status,
      children: children ?? this.children,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
